### ISSUE TITLE
Support for PhysicalMemory SVCs

### ISSUE THREAD

**Issue by fincs** [switchbrew/libnx#503]
In version 3.0 of the Horizon kernel, new memory management SVCs were added. These essentially implement a new, more flexible take on heap management. Newer official software is known to take advantage of this improved heap management system. As a result, it would be desirable to add support for PhysicalMemory SVCs to libnx so that homebrew can take advantage of it.

Adding support involves redesigning how libnx applications interact with the homebrew environment (specifically nx-hbloader), and also figuring out a way to fall back on older style svcSetHeapSize heap management on systems running older Horizon kernels. Also note that Mesosphère implements all newer Horizon kernel functionality regardless of the installed firmware version; and a future version of Atmosphère will unconditionally have Mesosphère enabled, therefore making this a non-issue.

**Comment by DCNick3** [comment]
Well, it seems that now Mesosphère is close to being unconditionally enabled.. Can you please look at this?

**Comment by Bulmanator** [comment]
Hello, I was wondering if there was an update to this?
As of Atmosphère 1.0.0 from what I can tell Mesosphère is no longer opt-out and thus unconditionally enabled.

It would be nice to be able to use generic virtual memory in homebrew applications to align with PC counterparts like VirtualAlloc/mmap (that said, I haven't wrapped my head around SVC call stuff, so this might not even be what I think it is)

**Comment by xfreshx** [comment]
Yes, having mmap is somestging I do need for a port of X-Ray game engine (open source version of famous S.T.A.L.K.E.R game engine). Would you provide any update please?

**Comment by lulle2007200** [comment]
Access to SvcMapPhysicalMemory from homebrew would be really useful for a project im working on. 

I'd like to (help to) implement support in libnx, but not quite sure about thr details. Especially how/if to handle cases where the new memory svc are not available, e.g. ancient HOS without Meso.

**Comment by diekleinekuh** [comment]
Having something like this available would be really nice. My use case is some code that first reserves a 4GB range of address space and then on demand hands out memory from that region relying on the fact that everything is in the same range so all addresses can be converted into a 32bit offset from the region start. 

Currently I have to allocate a certain memory block upfront that is large enough to fulfill all memory requests that use the offset addressing but leave enough room for other allocations that do not use this system.

**Comment by lulle2007200** [comment]
@fincs would you accept a PR to add support for the PhysicalMemory SVCs?
If yes, how should this be implemented while keeping compatibility with existing homebrew NROs?

Some (rough) ideas that came up when talking to other people wanting support:
1) Add a flag to all new NROs / NROs that want to use PhysicalMemory SVCs. In HBL, if set, do not allocate all available memory via SetHeapSize and instead let libnx handle heap allocations from the malloc family etc. via MapPhysicalMemory
2) Indicate support for PhysicalMemory SVCs via an environment entry set by HBL and keep the current behaviour (e.g. allocate all available memory in HBL). New NROs/NROs that want PhysicalMemory SVC support can check for the environment entry. The heap that was set up by HBL would have to be reset and the PhysicalMemory SVCs should be used instead.

HBL should set NPDM#SystemResourceSize to the maximum possible value.

As for why support would be useful:
- Required for porting dotnet runtime
- Useful for emulation or compatibility layers like vita2hos
- Useful for anything that requires mmap-like behaviour
